File: src/main/java/LibTMOA/models/structures/MecanumDirectives.java
/*
 * Copyright 2020 WinT 3794 (Manuel Díaz Rojo and Alexis Obed García Hernández)
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 *
 */

package LibTMOA.models.structures;

/**
 * ADT of Mecanum Directives [Vd, Td, Vt]
 */
public class MecanumDirectives {
    private double vd;
    private double td;
    private double vt;

    /**
     * Creates a MecanumDirectives object from Vd, Td and Vt.
     * @param vd Multiplicative Speed [0 - 1]
     * @param td Angle Direction [-Math.PI - Math.PI]
     * @param vt Change Speed [-1 - 1]
     */
    public MecanumDirectives(double vd, double td, double vt) {
        this.vd = vd;
        this.td = td;
        this.vt = vt;
    }

    /**
     * Creates a MecanumDirectives object from Vd and Td.
     * @param vd Multiplicative Speed [0 - 1]
     * @param td Angle Direction [-Math.PI - Math.PI]
     */
    public MecanumDirectives(double vd, double td) {
        this.vd = vd;
        this.td = td;
        this.vt = 0;
    }

    /**
     * Returns Multiplicative Speed
     * @return Multiplicative Speed [double]
     */
    public double getVd() {
        return vd;
    }

    /**
     * Returns Angle Direction
     * @return Angle Direction [double]
     */
    public double getTd() {
        return td;
    }

    /**
     * Returns Change Speed
     * @return Change Speed [double]
     */
    public double getVt() {
        return vt;
    }

    /**
     * Returns array with directives
     * @return MecanumDirectives as an array
     */
    public double[] getDirectives(){
        double[] directives = new double[3];
        directives[0] = vd;
        directives[1] = td;
        directives[2] = vt;
        return directives;
    }

    /**
     * Returns string with directives
     * @return MecanumDirectives as a string
     */
    @Override
    public String toString() {
        return "MecanumDirectives{" +
                "Vd=" + vd +
                ", Td=" + td +
                ", Vt=" + vt +
                '}';
    }
}

File: src/main/kotlin/org/wint3794/pathfollower/models/MecanumDirectives.kt
/*
 * Copyright 2020 WinT 3794 (Manuel Diaz Rojo and Alexis Obed Garcia Hernandez)
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 *
 */
package org.wint3794.pathfollower.models

/**
 * ADT of Mecanum Directives [Vd, Td, Vt]
 */
class MecanumDirectives {
    /**
     * Returns Desired Magnitude
     *
     * @return Desired Magnitude [double]
     */
    val vd: Double

    /**
     * Returns Desired Angle
     *
     * @return Desired Angle [double]
     */
    val td: Double

    /**
     * Returns Desired Rotation
     *
     * @return Desired Rotation [double]
     */
    val vt: Double

    /**
     * Creates a MecanumDirectives object from Vd, Td and Vt.
     *
     * @param vd Desired Magnitude [0 - 1]
     * @param td Desired Angle [0 - 2 * Math.PI]
     * @param vt Desired Rotation [-1 - 1]
     */
    constructor(vd: Double, td: Double, vt: Double) {
        this.vd = vd
        this.td = td
        this.vt = vt
    }

    /**
     * Creates a MecanumDirectives object from Vd and Td.
     *
     * @param vd Desired Magnitude [0 - 1]
     * @param td Desired Angle [0 - 2 * Math.PI]
     */
    constructor(vd: Double, td: Double) {
        this.vd = vd
        this.td = td
        vt = 0.0
    }

    /**
     * Returns array with directives
     *
     * @return MecanumDirectives as an array
     */
    val directives: DoubleArray
        get() {
            val directives = DoubleArray(3)
            directives[0] = vd
            directives[1] = td
            directives[2] = vt
            return directives
        }

    /**
     * Returns string with directives
     *
     * @return MecanumDirectives as a string
     */
    override fun toString(): String {
        return "MecanumDirectives{" +
                "Vd=" + vd +
                ", Td=" + td +
                ", Vt=" + vt +
                '}'
    }
}